src/adapter/src/coord/sequencer/inner/create_materialized_view.rs

```diff
@@ -540,8 +539,26 @@ impl Coordinator {
     ) -> Result<StageResult<Box<CreateMaterializedViewStage>>, AdapterError> {
         // Timestamp selection
         let id_bundle = dataflow_import_id_bundle(global_lir_plan.df_desc(), cluster_id);
-        let (read_holds, dataflow_as_of, storage_as_of, until) =
-            self.select_timestamps(id_bundle, refresh_schedule.as_ref())?;
+
+        let read_holds_owned;
+        let read_holds = if let Some(txn_reads) = self.txn_read_holds.get(session.conn_id()) {
+            // In some cases, for example when REFRESH is used, the preparatory
+            // stages will already have acquired ReadHolds, we can re-use those.
+            tracing::info!(session = %session.conn_id(), ?txn_reads, "using txn read holds for materialized view");
+
+            // We assume that earlier holds appear first in the txn holds.
+            let txn_reads = &txn_reads[0];
+            txn_reads
```

Comment:
Do these come from the read holds acquired in `resolve_mz_now_for_create_materialized_view`? If so do you know why we weren't doing something like this before?

Reply:
> Do these come from the read holds acquired in `resolve_mz_now_for_create_materialized_view`?

Yes, and also from `create_materialized_view_validate` (the `acquire_read_holds_auto_cleanup` call).

> so do you know why we weren't doing something like this before?

It's because the old `least_valid_read` call in `select_timestamps` would take into account the read holds acquired earlier. I think this is because it looks at the implied capabilities, which include read holds, because the read holds are installed by modifying the `ReadPolicy`, adding `ValidFrom`s to them, and the implied capabilities are based on `ReadPolicy`s.

Btw. you can try this on `main`: If you comment out the `acquire_read_holds` in both `resolve_mz_now_for_create_materialized_view` and `create_materialized_view_validate`, then the following fails:
```
create table t (x int);
INSERT INTO t VALUES (1);

CREATE MATERIALIZED VIEW mv
WITH (REFRESH = AT CREATION)
AS
(SELECT * FROM t)
UNION ALL
(SELECT CASE WHEN mz_unsafe.mz_sleep(3) IS NULL THEN 5 ELSE 8 END);
```
because the since has moved past the time of the `REFRESH = AT CREATION` by the time we get to `select_timestamps`.

Reply:
I now have all (most?) of the info paged back in:
- `resolve_mz_now_for_create_materialized_view` will store read holds in `txn_reads`, unless it's invoked as part of an `EXPLAIN`
- `create_materialized_view_validate` will put read holds in `txn_reads` for every time in the refresh schedule
- we don't have any read holds when the query is not timestamp/timeline dependent, for example `create materialized view mv as select 1`

As for why we didn't acquire read holds in this exact piece of code before: most moments were just using the sinces of collections and were assuming that they don't move along concurrently. That's somewhat brittle, and will actually not be true once I make the storage controller more concurrent (one of the end goals of this whole thing :sweat_smile:). Also, though, `select_timestamp` was previously acquiring new read holds, and I'm changing it to re-use those that we already have here.

Reply:
> As for why we didn't acquire read holds in this exact piece of code before: most moments were just using the sinces of collections and were assuming that they don't move along concurrently. That's somewhat brittle, and will actually not be true once I make the storage controller more concurrent (one of the end goals of this whole thing 😅).

I think the old `least_valid_read` _would_ take into account the read holds acquired in `resolve_mz_now_for_create_materialized_view` and `create_materialized_view_validate`, see my above comment.